File: database-driver-plugin/src/main/java/de/lino/database/database/nosql/csv/CSVDatabaseProvider.java
package de.lino.database.database.nosql.csv;

import de.lino.database.database.AbstractCachedDatabaseSection;
import de.lino.database.database.AbstractLazyDatabaseProvider;
import de.lino.database.database.SectionConfig;
import de.lino.database.database.auth.Credentials;
import de.lino.database.json.file.FileProvider;
import de.lino.database.database.DatabaseProvider;
import de.lino.database.database.DatabaseSection;
import org.jetbrains.annotations.NotNull;

import java.io.File;
import java.nio.file.Path;
import java.util.Objects;
import java.util.function.Consumer;

/**
 * The CSV-file-based {@link DatabaseProvider}: every {@link DatabaseSection} is one
 * {@code "<name>.csv"} file directly under {@link Credentials}'s {@code getFileRepository()},
 * via {@link CSVDatabaseSection}. Section lifecycle and caching live in
 * {@link AbstractLazyDatabaseProvider}; this class only supplies the file-level storage
 * operations - listing {@value #EXTENSION} files, constructing a {@link CSVDatabaseSection},
 * deleting a file.
 */
public class CSVDatabaseProvider extends AbstractLazyDatabaseProvider {

    /**
     * The file extension every section's file carries.
     */
    private static final String EXTENSION = ".csv";

    /**
     * The directory every section's CSV file lives directly under.
     */
    private final Path repository;

    /**
     * Discovers every existing {@value #EXTENSION} file directly under {@code credentials}'
     * file repository as a section name. Only names - no section objects, no file contents -
     * so construction cost is one directory listing, independent of how much data the
     * repository holds.
     *
     * @param credentials the login credentials, providing the file repository root this
     *                    database's sections live under
     */
    public CSVDatabaseProvider(@NotNull final Credentials credentials) {

        this.repository = Path.of(credentials.getFileRepository());

        this.reload();

    }

    @Override
    public void shutdown() {
    }

    /**
     * {@inheritDoc}
     * <p>
     * Lists the repository's {@value #EXTENSION} files, (re-)creating the repository root
     * first so a fresh installation starts from an existing, empty directory rather than
     * failing to list a missing one; each file name minus the extension is one section name.
     */
    @Override
    protected void discoverNames(@NotNull Consumer<String> consumer) {

        FileProvider.getInstance().createDirectory(this.repository);

        final File[] files = this.repository.toFile().listFiles((directory, fileName) -> fileName.endsWith(EXTENSION));

        for (final File file : Objects.requireNonNull(files)) {
            consumer.accept(file.getName().substring(0, file.getName().length() - EXTENSION.length()));
        }

    }

    @Override
    protected AbstractCachedDatabaseSection constructSection(@NotNull String name, @NotNull SectionConfig config) {
        return new CSVDatabaseSection(name, this.repository.resolve(name + EXTENSION), config);
    }

    @Override
    protected void dropSectionRemote(@NotNull String name) {
        FileProvider.getInstance().deleteFile(this.repository.resolve(name + EXTENSION));
    }

}

File: python/database-driver-plugin/src/database_driver/plugin/database/nosql/csv/csv_database_provider.py
"""Mirror of ``de.lino.database.database.nosql.csv.CSVDatabaseProvider``."""

from __future__ import annotations

from collections.abc import Callable
from pathlib import Path

from database_driver.api.database.auth.credentials import Credentials
from database_driver.api.database.section_config import SectionConfig
from database_driver.api.json.file.file_provider import FileProvider
from database_driver.plugin.database.abstract_cached_database_section import AbstractCachedDatabaseSection
from database_driver.plugin.database.abstract_lazy_database_provider import AbstractLazyDatabaseProvider
from database_driver.plugin.database.nosql.csv.csv_database_section import CSVDatabaseSection

_EXTENSION = ".csv"
"""The file extension every section's file carries."""


class CSVDatabaseProvider(AbstractLazyDatabaseProvider):
    """The CSV-file-based ``DatabaseProvider``: every section is one ``<name>.csv`` file
    directly under the credentials' file repository, via ``CSVDatabaseSection``. Section
    lifecycle and caching live in ``AbstractLazyDatabaseProvider``; this class only
    supplies the file-level storage operations - listing ``.csv`` files, constructing a
    section, deleting a file. Discovers only names at construction, so construction cost
    is one directory listing, independent of how much data the repository holds."""

    def __init__(self, credentials: Credentials) -> None:
        super().__init__()
        self.repository = Path(credentials.file_repository)
        self.reload()

    def shutdown(self) -> None:
        pass

    def discover_names(self, consumer: Callable[[str], None]) -> None:
        """Lists the repository's ``.csv`` files, (re-)creating the repository root first
        so a fresh installation starts from an existing, empty directory rather than
        failing to list a missing one; each file name minus the extension is one section
        name."""
        provider = FileProvider.get_instance()
        assert provider is not None
        provider.create_directory(self.repository)

        for path in self.repository.glob(f"*{_EXTENSION}"):
            if path.is_file():
                consumer(path.name[: -len(_EXTENSION)])

    def construct_section(self, name: str, config: SectionConfig) -> AbstractCachedDatabaseSection:
        return CSVDatabaseSection(name, self.repository / f"{name}{_EXTENSION}", config)

    def drop_section_remote(self, name: str) -> None:
        provider = FileProvider.get_instance()
        assert provider is not None
        provider.delete_file(self.repository / f"{name}{_EXTENSION}")
